ADR/0018-apps-continuous-perf-testing.md

```diff
@@ -0,0 +1,72 @@
+# 18. Continuous Performance Testing (CPT) of Apps in Stonesoup
+
+Date: 2023-03-10
+
+## Status
+
+In consideration
+
+## Context
+
+In general, performance testing is just another form of testing that helps application teams to ensure there are no regressions in their code and that their application behaves as expected.
+
+The IntegrationTestScenario pipelines in StoneSoup are not suitable for full-blown performance and scale testing (that usually takes lots of time and involves human analysis), but are a good place for a quick and small-scale regression test.
+
+What makes performance testing different from functional testing is it is harder to decide if the test passed or failed as every performance test tests different aspects of the application and so it expects different performance, different metrics, different thresholds. Furthermore, even if the measured performance of the app under test did not change, there might be some significant resource usage by the application which we want to cause the test to be marked as failed.
+
+The approach to make this pass/fail detection possible proposed here, is to use historical results as a benchmark – not only actual results of the performance test, but also monitoring data about the resource usage, etc.
+
+> Example: Imagine you develop a web e-shop application that uses PostgreSQL data backend. Your perf test is browsing through a goods catalog and measures latency of pages. When you want to decide on pass or fail result of this test, you need to check that metrics like below are aligned to previous results of a same test with same configuration:
+>
+> - Average page load latency.
+> - Backend service CPU and memory usage...
+> - PostgreSQL CPU and memory usage…
+> - Internal PostgreSQL metrics like number of commits during the test or average number of database sessions during the test…
+>
+> If any of these metrics does not align to historical results, we mark this test result as a failure.
+
+And even if some test fails, the application team should be able to review the data and change the test definition, so next time a new test result is being evaluated based on the historical results, this new result will be one of them.
+
+The algorithm that says if the current test passes or fails when compared to historical results can vary. It can be:
+
+- Checking minimum and maximum of the metric in historical data and making sure the new result is in this range.
+- Using a floating average of historical data and setting a safe range around it and making sure the new result is in that range.
+- Some machine learning thingy.
+
+Goal of this ADR is to propose a way for this kind of testing to be implemented in Stonesoup (feature [STONE-679](https://issues.redhat.com/browse/STONE-679)). Even if it would not deprecate full-blown performance and scale testing, having a release gate with some short and small scale performance test is desirable for many application teams.
```

Comment:
What are the differences between the proposed approach and "full-blown performance and scale testing"? Is there a comparison between them to highlight the benefits of one over the other? How is the current proposal not sufficient for a "full blown" test?

Reply:
Hello! I think this depends on how are you doing performance and scale testing. In my expirience it is usually far from being fully automated as it involves lots of iterations and human analysis of resutls. Also some testing can be very expensive and is not happening often. The testing we are trying to propose here is just an fully automated performance regression check. I have added some more explanation to `The IntegrationTestScenario pipelines in StoneSoup are not suitable...` paragraph about as well.

Reply:
Hope it is OK to resolve this now. Please reopen if there is something to discuss.